### ISSUE TITLE
Menu dropdown is not always on top in click event order

### ISSUE THREAD

**Issue by rafaqz** [MakieOrg/Makie.jl#3032]
If a menu opens over an axis its possible for the axis click events to come first, meaning the menu can't be selected.

Maybe menu dropdowns should always get click events first?

**Comment by ffreyer** [comment]
Menu should already have priority over Axis. Do you have an example when it does not?

**Comment by rafaqz** [comment]
I do but its horribly complicated. How is priority determined?

**Comment by ffreyer** [comment]
It's a hardcoded value. For Menu:
https://github.com/MakieOrg/Makie.jl/blob/77823ff524415f080390240c18d11ae6403e942d/src/makielayout/blocks/menu.jl#L224

For Axis I believe it falls onto the default of `addmouseevents!()` which is 1:
https://github.com/MakieOrg/Makie.jl/blob/77823ff524415f080390240c18d11ae6403e942d/src/makielayout/mousestatemachine.jl#L115

Is it lag related or render order related (i.e. Axis drawing over the dropdown)?

**Comment by rafaqz** [comment]
Ok probably its priority related then. MakieDraw.jl uses high priority to make sure it captures the mouse over the underlying plot.

Maybe we need some place to list what objects use what priorities. Is it in the docs?

Also seems a little fragile that user defined event priorities can break the menu.

**Comment by ffreyer** [comment]
Yea should be. The maximum priority used to be 127 iirc, and I picked half of that so we still had room to move up. Since it got moved to Observables we switched to full Int64's for priority, so we can and probably should just pump this to some bigger number.

Making it less fragile seems difficult to me. We could put the Menu at maximum priority, but then what if someone wanted to implement something like a help screen, or a pause menu or a screenshot interaction that lets you do a rectangle selection while freezing everything else? Those things would need to have even higher priority.
Something we could think about is adding a function that tells you whether a scene is visible at a given position, so that you can skip interaction when that is not the case. We'd need to make sure that scene order is consistent across backends first though. We'd also need to make less scenes transparent, otherwise this would be rather useless.

**Comment by rafaqz** [comment]
I agree something a lot higher than `64` is a good idea. Few things will want to be above an open dropdown menu, but you're right they will exist.

A way to know if an object is visible at a specific click location could be useful too.

**Comment by ffreyer** [comment]
I guess we should just define an enum with a couple named priorities and use those. Since the priority stuff moved to Observables.jl we got more space to use (full Int64) so the number may have changed.
From what I remember we need:
- `typemax(Int64)` for listeners that should always run and never Consume (e.g. keyboardbutton setting keyboardstate)
- `typemax(Int64) - 1` would probably be good to add for "block everything" listeners, e.g. the tick overwrite in record
- high priority range for Widgets that may overlap with other widgets (e.g. Menu)
- medium priority for spatially isolated Widgets & Axis 
- `0` as default

Adding this shouldn't be hard, but will require some searching for all the relevant event listeners. Should be save to ignore backends though.

**Comment by rafaqz** [comment]
OMG I forgot I even made this issue and we worked out the answer, I had this problem again in MakieDraw.jl last week.

That makes me think named priorities would help, these magic numbers are easily forgotten.
